File: platforms/ios/Dezel/Dezel/Modules/Device/Device.swift
import AudioToolbox

/**
 * @class Device
 * @since 0.4.0
 * @hidden
 */
open class Device: JavaScriptClass {

	//--------------------------------------------------------------------------
	// MARK: JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsGet_name
	 * @since 0.4.0
	 * @hidden
	 */
	@objc open func jsGet_uuid(callback: JavaScriptGetterCallback) {

		var uuid = UserDefaults.standard.string(forKey: "dezel.device.uuid")
		if (uuid == nil) {
			uuid = UUID().uuidString
			UserDefaults.standard.set(uuid, forKey: "dezel.device.uuid")
		}

		callback.returns(string: uuid!)
	}

	//--------------------------------------------------------------------------
	// MARK: JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_sound
	 * @since 0.4.0
	 * @hidden
	 */
	@objc open func jsFunction_sound(callback: JavaScriptFunctionCallback) {

		switch (callback.argument(0).string) {

			case "shutter":
				AudioServicesPlaySystemSound(SystemSoundID(1108))
				return

			default:
				break
		}

		AudioServicesPlaySystemSound(SystemSoundID(callback.argument(0).number.int()))
	}

	/**
	 * @method jsFunction_vibrate
	 * @since 0.4.0
	 * @hidden
	 */
	@objc open func jsFunction_vibrate(callback: JavaScriptFunctionCallback) {

		if (callback.arguments == 0) {
			AudioServicesPlaySystemSound(SystemSoundID(4095))
			return
		}

		if (callback.arguments == 1) {

			switch (callback.argument(0).string) {

				case "peek":
					AudioServicesPlaySystemSound(SystemSoundID(1519))
					return

				case "pop":
					AudioServicesPlaySystemSound(SystemSoundID(1520))
					return

				case "cancel":
					AudioServicesPlaySystemSound(SystemSoundID(1521))
					return

				case "retry":
					AudioServicesPlaySystemSound(SystemSoundID(1102))
					return

				case "failed":
					AudioServicesPlaySystemSound(SystemSoundID(1107))
					return

				default:
					AudioServicesPlaySystemSound(SystemSoundID(4095))
					break
			}
		}

		AudioServicesPlaySystemSound(SystemSoundID(callback.argument(0).number.int()))
	}
}

File: platforms/android/Dezel/dezel/src/main/java/ca/logaritm/dezel/modules/device/Device.kt
package ca.logaritm.dezel.modules.device

import android.content.Context.VIBRATOR_SERVICE
import android.content.SharedPreferences
import android.media.MediaActionSound
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import ca.logaritm.dezel.core.JavaScriptClass
import ca.logaritm.dezel.core.JavaScriptContext
import ca.logaritm.dezel.core.JavaScriptFunctionCallback
import ca.logaritm.dezel.core.JavaScriptGetterCallback
import java.util.*

/**
 * @class Device
 * @since 0.4.0
 * @hidden
 */
open class Device(context: JavaScriptContext): JavaScriptClass(context) {

	/**
	 * @property preferences
	 * @since 0.1.0
	 * @hidden
	 */
	private val preferences: SharedPreferences = this.context.application.getSharedPreferences("dezel.device.uuid", android.content.Context.MODE_PRIVATE)

	//--------------------------------------------------------------------------
	// JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsGet_name
	 * @since 0.4.0
	 * @hidden
	 */
	open fun jsGet_uuid(callback: JavaScriptGetterCallback) {

		var value = this.preferences.getString("dezel.device.uuid", null)
		if (value == null) {

			val editor = this.preferences.edit()
			if (editor == null) {
				return
			}

			value = UUID.randomUUID().toString()

			editor.putString("dezel.device.uuid", value)
			editor.apply()
		}

		callback.returns(value)
	}

	//--------------------------------------------------------------------------
	// JS Functions
	//--------------------------------------------------------------------------

	/**
	 * @method jsFunction_sound
	 * @since 0.4.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_sound(callback: JavaScriptFunctionCallback) {

		val value = callback.argument(0)
		if (value.isNull ||
			value.isUndefined) {
			return
		}

		if (value.isString) {

			when (value.string) {

				"shutter" -> {
					MediaActionSound().play(MediaActionSound.SHUTTER_CLICK)
					return
				}

			}

			return
		}

		MediaActionSound().play(value.number.toInt())
	}

	/**
	 * @method jsFunction_vibrate
	 * @since 0.4.0
	 * @hidden
	 */
	@Suppress("unused")
	open fun jsFunction_vibrate(callback: JavaScriptFunctionCallback) {

		val vibrator: Vibrator

		try {
			vibrator = this.context.application.getSystemService(VIBRATOR_SERVICE) as Vibrator
		} catch (e: Exception) {
			return
		}

		if (callback.arguments == 0) {

			if (Build.VERSION.SDK_INT >= 26) {
				vibrator.vibrate(VibrationEffect.createOneShot(150, VibrationEffect.DEFAULT_AMPLITUDE))
			} else {
				vibrator.vibrate(150)
			}

			return
		}
	}
}
